cabal-install/src/Distribution/Client/ScriptUtils.hs

```diff
@@ -0,0 +1,342 @@
+    cliConfig = commandLineFlagsToProjectConfig globalFlags flags mempty
+    globalConfigFlag = projectConfigConfigFile (projectConfigShared cliConfig)
+    defaultTarget = [TargetPackage TargetExplicitNamed [fakePackageId] Nothing]
+
+    with = do
+      ctx <- establishProjectBaseContext verbosity cliConfig OtherCommand
+      return (ProjectContext, ctx)
+    without mkDir globalConfig = do
+      distDirLayout <- establishDummyDistDirLayout verbosity (globalConfig <> cliConfig) =<< mkDir
+      ctx <- establishDummyProjectBaseContext verbosity (globalConfig <> cliConfig) distDirLayout [] OtherCommand
+      return (GlobalContext, ctx)
+    scriptOrError script err = do
+      exists <- doesFileExist script
+      if exists then do
+        -- In the script case we always want a dummy context even when ignoreProject is False
+        let mkCacheDir = ensureScriptCacheDirectory verbosity cachePrefix script
+        (_, ctx) <- withProjectOrGlobalConfig verbosity (Flag True) globalConfigFlag with (without mkCacheDir)
+
+        let projectRoot = distProjectRootDirectory $ distDirLayout ctx
+        writeFile (projectRoot </> "scriptlocation") =<< canonicalizePath script
+
+        (executable, contents) <- readScriptBlockFromScript verbosity =<< BS.readFile script
+
+        let executable' = executable & L.buildInfo . L.defaultLanguage %~ maybe (Just Haskell2010) Just
+        return (ScriptContext script executable' contents, ctx, defaultTarget)
+      else reportTargetSelectorProblems verbosity err
+
+withTemporaryTempDirectory :: (IO FilePath -> IO a) -> IO a
+withTemporaryTempDirectory act = newEmptyMVar >>= \m -> bracket (getMkTmp m) (rmTmp m) act
+  where
+    -- We return an (IO Filepath) instead of a FilePath for two reasons:
+    -- 1) To give the consumer the discretion to not create the tmpDir,
+    --    but still grantee that it's deleted if they do create it
+    -- 2) Because the path returned by createTempDirectory is not predicable
+    getMkTmp m = return $ do
+      tmpDir <- getTemporaryDirectory >>= flip createTempDirectory "cabal-repl."
+      putMVar m tmpDir
+      return tmpDir
+    rmTmp m _ = tryTakeMVar m >>= maybe (return ()) (handleDoesNotExist () . removeDirectoryRecursive)
+
+-- | Add the 'SourcePackage' to the context and use it to write a fake-package.cabal file.
+updateContextAndWriteProjectFile :: ProjectBaseContext -> SourcePackage (PackageLocation (Maybe FilePath)) -> IO ProjectBaseContext
+updateContextAndWriteProjectFile ctx srcPkg = do
+  let projectRoot = distProjectRootDirectory $ distDirLayout ctx
+      projectFile = projectRoot </> "fake-package.cabal"
+      writeProjectFile = writeGenericPackageDescription (projectRoot </> "fake-package.cabal") (srcpkgDescription srcPkg)
+  projectFileExists <- doesFileExist projectFile
+  -- TODO This if is here to prevent reconfiguration of cached repl packages.
+  -- It's worth investigating why it's need in the first place.
```

Comment:
```suggestion
  -- TODO This is here to prevent reconfiguration of cached repl packages.
  -- It's worth investigating why it's needed in the first place.
```